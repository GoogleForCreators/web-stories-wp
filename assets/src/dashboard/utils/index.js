/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { default as clamp } from './clamp';
export { default as getCurrentYAxis } from './getCurrentYAxis';
export { default as groupBy } from './groupBy';
export { default as keyboardOnlyOutline } from './keyboardOnlyOutline';
export { default as memoize } from './memoize';
export { default as throttleToAnimationFrame } from './throttleToAnimationFrame';
export { default as useDashboardResultsLabel } from './useDashboardResultsLabel';
export { default as useFocusOut } from './useFocusOut';
export { default as useGridViewKeys } from './useGridViewKeys';
export {
  default as usePagePreviewSize,
  getPagePreviewHeights,
} from './usePagePreviewSize';
export { default as useStoryView } from './useStoryView';
export { default as useTemplateView } from './useTemplateView';
export { default as validateGoogleAnalyticsIdFormat } from './validateGoogleAnalyticsIdFormat';

export { default as addQueryArgs } from '../../edit-story/utils/addQueryArgs';
export { default as getStoryPropsToSave } from '../../edit-story/app/story/utils/getStoryPropsToSave';
export { useKeyDownEffect } from '../../edit-story/components/keyboard';
export {
  focusOnPage,
  getArrowDir,
  getGridColumnAndRowCount,
  getRow,
  getColumn,
  getIndex,
} from '../../edit-story/components/canvas/gridview/useGridViewKeys';
export { default as useResizeEffect } from '../../edit-story/utils/useResizeEffect';
// TODO use these shared utils to structure image resources mirroring editor
export { default as getTypeFromMime } from '../../edit-story/app/media/utils/getTypeFromMime';
export { default as getResourceFromLocalFile } from '../../edit-story/app/media/utils/getResourceFromLocalFile';
export {
  createContext,
  identity,
  useContextSelector,
} from '../../edit-story/utils/context';
