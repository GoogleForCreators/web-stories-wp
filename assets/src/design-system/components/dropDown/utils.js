/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const getOptions = (groups) => {
  const isNested = groups.some((group) => Array.isArray(group?.options));
  if (isNested) {
    return groups.reduce((prev, current) => {
      // Determine if this nested group contains options as objects that contain a valid value key.
      const onlyValidOptions =
        current?.options &&
        current.options.filter(
          (option) => typeof option === 'object' && option?.value?.toString()
        );
      // If there aren't any valid options to return, just return the previously reduced array as is
      if (!Array.isArray(onlyValidOptions)) {
        return prev;
      }

      // Otherwise add the validated options to the reduced array as a new object containing group and label
      return prev.concat({
        group: onlyValidOptions,
        label: current?.label,
      });
    }, []);
  } else {
    // Double check that all data in groups should be treated as options to sanitize data
    const onlyValidOptions = groups.filter(
      (option) => typeof option === 'object' && option?.value
    );

    if (!Array.isArray(onlyValidOptions) || onlyValidOptions.length === 0) {
      return [];
    }
    return [{ group: onlyValidOptions }];
  }
};

export const getInset = (groups, i, j) =>
  groups
    .slice(0, i)
    .map(({ group }) => group.length)
    .reduce((a, b) => a + b, 0) + j;
