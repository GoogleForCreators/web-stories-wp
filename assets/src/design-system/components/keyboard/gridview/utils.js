/*
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// @todo: provide a cleaner `focusFirst()` API and takes into account
// `tabIndex`, `disabled`, links, buttons, inputs, etc.
export function focusOnPage(page) {
  const button = page?.querySelector('button');
  if (button) {
    button.focus();
  }
}

export function getArrowDir(key, pos, neg, isRTL) {
  const rtlDir = isRTL ? -1 : 1;
  if (key === pos) {
    return rtlDir;
  }
  if (key === neg) {
    return -1 * rtlDir;
  }
  return 0;
}

export function getGridColumnAndRowCount(grid, pageCount) {
  let columns = 0;
  let prevX;
  for (const el of grid.children) {
    const { x } = el.getBoundingClientRect();
    if (prevX != null && x < prevX) {
      break;
    }
    prevX = x;
    columns++;
  }

  const rows = Math.ceil(pageCount / columns);

  return { rows, columns };
}

// will return a 1 based index
export function getRow(index, numColumns) {
  return Math.ceil((index + 1) / numColumns);
}

// will return a 1 based index
export function getColumn(index, numColumns) {
  return (index % numColumns) + 1;
}

// will return a 0 based index
export function getIndex({ row, column, numRows, numColumns, numItems }) {
  const isOutOfBounds =
    row > numRows || row <= 0 || column > numColumns || column <= 0;

  if (isOutOfBounds) {
    return -1;
  }

  const index = numColumns * (row - 1) + (column - 1);

  // If the index is greater than or equal to numItems default to the last index.
  // This handles the case when we press ArrowDown and there is another row, but
  // the column below is empty. It will default to the last item in the list.
  return index >= numItems ? numItems - 1 : index;
}
