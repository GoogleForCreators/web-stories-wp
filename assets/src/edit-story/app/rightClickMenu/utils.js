/*
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import objectPick from '../../utils/objectPick';
import { ELEMENT_TYPES } from '../story';

const elementTypes = Object.values(ELEMENT_TYPES);

const MEDIA_STYLE_PROPERTIES = [
  'border',
  'borderRadius',
  'mask',
  'opacity',
  'overlay',
];

/**
 * Extracts the styles from an element based off of the element type.
 *
 * @param {Object} element the element
 * @return {Object} an object with a subset of properties of the original element. The properties
 * should all be properties that define styles on that element.
 */
export const getElementStyles = (element) => {
  if (!element || !element?.type || !elementTypes.includes(element.type)) {
    return null;
  }

  let properties = [];

  switch (element?.type) {
    case ELEMENT_TYPES.BACKGROUND:
      properties = ['backgroundColor', 'mask', 'opacity', 'overlay'];
      break;
    case ELEMENT_TYPES.GIF:
      properties = MEDIA_STYLE_PROPERTIES;
      break;
    case ELEMENT_TYPES.IMAGE:
      properties = MEDIA_STYLE_PROPERTIES;
      break;
    case ELEMENT_TYPES.SHAPE:
      properties = ['backgroundColor', 'border', 'borderRadius', 'opacity'];
      break;
    case ELEMENT_TYPES.TEXT:
      properties = [
        'backgroundColor',
        'backgroundTextMode',
        'border',
        'borderRadius',
        'font',
        'fontSize',
        'fontWeight',
        'lineHeight',
        'opacity',
        'padding',
        'textAlign',
      ];
      break;
    case ELEMENT_TYPES.VIDEO:
      properties = MEDIA_STYLE_PROPERTIES;
      break;
    default:
      break;
  }

  return objectPick(element, properties);
};
