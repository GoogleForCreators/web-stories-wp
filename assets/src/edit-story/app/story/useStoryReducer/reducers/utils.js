/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { LAYER_DIRECTIONS } from '../../../../constants';
import { ELEMENT_RESERVED_PROPERTIES } from '../types';
import objectWithout from '../../../../utils/objectWithout';
export { objectWithout };

export function intersect(first, ...rest) {
  if (!first || !rest || rest.length === 0) {
    return first;
  }

  return rest.reduce(
    (intersection, list) =>
      intersection.filter((value) => list.includes(value)),
    first
  );
}

export function isInsideRange(index, start, end) {
  return index >= start && index <= end;
}

export function moveArrayElement(array, oldPosition, newPosition) {
  // First remove from list.
  const element = array[oldPosition];
  const arrayWithoutElement = [
    ...array.slice(0, oldPosition),
    ...array.slice(oldPosition + 1),
  ];

  // Then re-insert at the right point
  return [
    ...arrayWithoutElement.slice(0, newPosition),
    element,
    ...arrayWithoutElement.slice(newPosition),
  ];
}

export function getAbsolutePosition({
  currentPosition,
  minPosition,
  maxPosition,
  desiredPosition,
}) {
  if (typeof desiredPosition === 'number') {
    return Math.min(maxPosition, Math.max(minPosition, desiredPosition));
  }

  if (typeof desiredPosition !== 'string') {
    return currentPosition;
  }

  switch (desiredPosition) {
    case LAYER_DIRECTIONS.FRONT:
      return maxPosition;
    case LAYER_DIRECTIONS.BACK:
      return minPosition;
    case LAYER_DIRECTIONS.FORWARD:
      return Math.min(maxPosition, currentPosition + 1);
    case LAYER_DIRECTIONS.BACKWARD:
      return Math.max(minPosition, currentPosition - 1);
    default:
      return currentPosition;
  }
}

export function updateElementWithUpdater(element, properties) {
  const updater =
    typeof properties === 'function' ? properties(element) : properties;
  const allowedProperties = objectWithout(updater, ELEMENT_RESERVED_PROPERTIES);
  if (Object.keys(allowedProperties).length === 0) {
    return element;
  }
  return { ...element, ...allowedProperties };
}

export function removeAnimationsWithElementIds(animations, ids) {
  return (animations || []).reduce((accum, animation) => {
    if (ids.some((id) => animation.targets?.includes(id))) {
      return accum;
    }
    return [...accum, animation];
  }, []);
}

export function updateAnimations(oldAnimations, animationUpdates) {
  const newAnimations = oldAnimations.reduce((animations, animation) => {
    const updatedAnimation = animationUpdates[animation.id];

    // remove animation from lookup
    delete animationUpdates[animation.id];

    if (updatedAnimation?.delete) {
      // delete animation
      return animations;
    } else if (updatedAnimation) {
      // update animation
      return [...animations, updatedAnimation];
    } else {
      // No updates
      return [...animations, animation];
    }
  }, []);

  // add animations
  Object.values(animationUpdates).forEach((a) => newAnimations.push(a));

  return newAnimations;
}

/**
 * @typedef {{ id: string, [prop: string]: any}} Entry
 * @property {string} id - unique identifier for entry
 */

/**
 * Remove duplicate entries. Uses last instance if
 * multiple entries share the same id.
 *
 * @param {Array<Entry>} entries - set of entries with possible duplicate Ids
 * @return {Array<Entry>} New set of entries with only unique Ids
 */
export function removeDuplicates(entries = []) {
  // Use only last of multiple elements with same id by turning into an object and getting the values.
  return Object.values(
    Object.fromEntries(entries.map((entry) => [entry.id, entry]))
  );
}

/**
 * Takes to sets of entries and returns unique entries
 * (keying on id) of right set not present in left set.
 *
 * @param {Array<Entry>} left - base set of entries
 * @param {Array<Entry>} right - new entries
 * @return {Array<Entry>} - right exclusion of sets set
 */
export function exclusion(left = [], right = []) {
  const rightSet = removeDuplicates(right);
  const leftJoinKeys = left.map(({ id }) => id);
  return rightSet.filter(({ id }) => !leftJoinKeys.includes(id));
}
