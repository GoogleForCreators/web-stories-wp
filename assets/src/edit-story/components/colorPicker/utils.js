/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// These functions are used as wrapper to allow test library to simulate events
// correctly even though pointer events are not supported by jsdom library

export function getPageX(evt) {
  return evt.pageX || evt.clientX;
}

export function getPageY(evt) {
  return evt.pageY || evt.clientY;
}

export function setPointerCapture(evt) {
  // Ignore reason: PointerEvents don't work in JSDOM
  // istanbul ignore if
  if (evt.target.setPointerCapture) {
    evt.target.setPointerCapture(evt.pointerId);
  }
}

export function releasePointerCapture(evt) {
  // Ignore reason: PointerEvents don't work in JSDOM
  // istanbul ignore if
  if (evt.target.releasePointerCapture) {
    evt.target.releasePointerCapture(evt.pointerId);
  }
}
