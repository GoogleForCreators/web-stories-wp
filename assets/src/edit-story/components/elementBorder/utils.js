/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { canMaskHaveBorder } from '../../masks';
import { BORDER_POSITION } from '../../constants';

export function shouldDisplayBorder(element) {
  const { border } = element;
  if (!border) {
    return false;
  }
  const { left, top, right, bottom, color } = border;
  // If we have no color, let's short-circuit.
  if (!color) {
    return false;
  }
  // If we have no border set either, let's short-circuit.
  if (!left && !top && !right && !bottom) {
    return false;
  }

  return canMaskHaveBorder(element);
}

export function getBorderPositionCSS({
  left,
  top,
  right,
  bottom,
  position,
  width = '100%',
  height = '100%',
  posTop = '0px',
  posLeft = '0px',
  skipOutsideBorder = true,
}) {
  if (BORDER_POSITION.CENTER === position) {
    return {
      top: `calc(${posTop} - ${top / 2}px)`,
      height: `calc(${height} + ${(top + bottom) / 2}px)`,
      left: `calc(${posLeft} - ${left / 2}px)`,
      width: `calc(${width} + ${(left + right) / 2}px)`,
    };
  }
  if (isOutsideBorder({ position }) && !skipOutsideBorder) {
    return {
      left: `calc(${posLeft} - ${left}px)`,
      top: `calc(${posTop} - ${top}px)`,
      width: `calc(${width} + ${left + right}px)`,
      height: `calc(${height} + ${top + bottom}px)`,
    };
  }
  return '';
}

export function getBorderStyle({
  color: rawColor,
  left,
  top,
  right,
  bottom,
  position,
  borderRadius,
  opacity = 100,
  skipOutsideBorder = true,
}) {
  const {
    color: { r, g, b, a },
  } = rawColor;
  // In case of inner/center border we need to adjust the opacity based on the layer opacity, too.
  // Outside border already has opacity applied since it's added to the element directly.
  opacity = opacity / 100;
  const adjustedBorderOpacity = a !== undefined ? opacity * a : opacity;
  const color = isOutsideBorder({ position })
    ? `rgba(${r},${g},${b},${a === undefined ? 1 : a})`
    : `rgba(${r},${g},${b},${adjustedBorderOpacity})`;

  // We're making the border-width responsive just for the preview,
  // since the calculation is not 100% precise here, we're opting to the safe side by rounding the widths up
  // as opposed to having potential margin between the border and the element.
  // When not in preview, the rounding doesn't have any effect.
  const borderWidth = `${Math.ceil(top)}px ${Math.ceil(right)}px ${Math.ceil(
    bottom
  )}px ${Math.ceil(left)}px`;
  const borderStyle = {
    ...getBorderPositionCSS({
      left,
      top,
      right,
      bottom,
      position,
      skipOutsideBorder,
    }),
    borderWidth,
    borderColor: color,
    borderStyle: 'solid',
    borderRadius: borderRadius
      ? `${borderRadius.topLeft}px ${borderRadius.topRight}px ${borderRadius.bottomRight}px ${borderRadius.bottomLeft}px`
      : null,
  };
  return {
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    width: '100%',
    height: '100%',
    position: 'absolute',
    ...borderStyle,
  };
}

export function isOutsideBorder(border) {
  return border?.position === BORDER_POSITION.OUTSIDE;
}

export function getInnerRadius(outerRadius, oneSide, otherSide) {
  return outerRadius - Math.min(outerRadius, Math.max(oneSide, otherSide)) / 2;
}

export function getBorderRadius({ borderRadius, border }) {
  if (!borderRadius) {
    return {};
  }
  if (border?.position === BORDER_POSITION.CENTER) {
    const radii = [
      getInnerRadius(borderRadius.topLeft, border.top, border.left),
      getInnerRadius(borderRadius.topRight, border.top, border.right),
      getInnerRadius(borderRadius.bottomRight, border.bottom, border.right),
      getInnerRadius(borderRadius.bottomLeft, border.bottom, border.left),
    ];
    return {
      borderRadius: radii.map((radius) => `${radius}px`).join(' '),
    };
  }
  return {
    borderRadius: `${borderRadius.topLeft}px ${borderRadius.topRight}px ${borderRadius.bottomRight}px ${borderRadius.bottomLeft}px`,
  };
}
