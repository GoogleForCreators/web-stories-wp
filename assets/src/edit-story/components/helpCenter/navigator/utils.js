/*
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * External dependencies
 */
import ResizeObserver from 'resize-observer-polyfill';

/**
 * Internal dependencies
 */
import { BEZIER } from '../../../../animation';

export function removeInnerElementFromLayoutFlow(innerEl, outerEl) {
  if (!innerEl || !outerEl) {
    return () => {};
  }

  // Explicitly set height of container, and disconnect
  // inner content from layout flow.
  const { height } = innerEl.getBoundingClientRect();
  outerEl.style.height = `${height}px`;
  innerEl.style.position = 'absolute';
  innerEl.style.bottom = 0;
  innerEl.style.content = 'contain';

  // Applying transition on separate frame as a
  // change to an animated property prevents any
  // content flash. Only way to guarentee the frame
  // after styles get applied is to do 2 frames from now.
  let id = requestAnimationFrame(() => {
    id = requestAnimationFrame(() => {
      outerEl.style.transition = `0.3s height ${BEZIER.default}`;
    });
  });
  return () => cancelAnimationFrame(id);
}

export function syncOuterHeightWithInner(innerEl, outerEl) {
  if (!innerEl) {
    return () => {};
  }
  const observer = new ResizeObserver((entries) => {
    const measureEl = entries?.[0];
    if (!measureEl || !outerEl) {
      return;
    }
    outerEl.style.height = `${measureEl.contentRect.height}px`;
  });
  observer.observe(innerEl);
  return () => observer.unobserve(innerEl);
}
