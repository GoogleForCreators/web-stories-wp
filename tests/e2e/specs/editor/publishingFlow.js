/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import {
  getEditedPostContent,
  publishPostWithPrePublishChecksDisabled,
  arePrePublishChecksEnabled,
  disablePrePublishChecks,
  enablePrePublishChecks,
} from '@wordpress/e2e-test-utils';

/**
 * Internal dependencies
 */
import { createNewStory, addRequestInterception } from '../../utils';

describe('Publishing Flow', () => {
  let stopRequestInterception;

  beforeAll(async () => {
    await page.setRequestInterception(true);
    stopRequestInterception = addRequestInterception((request) => {
      if (request.url().startsWith('https://cdn.ampproject.org/')) {
        request.respond({
          status: 200,
          body: '',
        });
      } else {
        request.continue();
      }
    });
  });

  afterAll(async () => {
    await page.setRequestInterception(false);
    stopRequestInterception();
  });

  it('should guide me towards creating a new post to embed my story', async () => {
    await createNewStory();

    await expect(page).toMatchElement('input[placeholder="Add title"]');
    await page.type('input[placeholder="Add title"]', 'Publishing Flow Test');

    // Publish story.
    await expect(page).toClick('button', { text: 'Publish' });

    await expect(page).toMatchElement('button', { text: 'Dismiss' });

    // Create new post and embed story.
    await expect(page).toClick('a', { text: 'Add to new post' });
    await page.waitForNavigation();

    // Disable Gutenberg's Welcome Guide if existing.
    const isWelcomeGuideActive = await page.evaluate(() =>
      wp.data.select('core/edit-post').isFeatureActive('welcomeGuide')
    );

    //eslint-disable-next-line jest/no-if
    if (isWelcomeGuideActive) {
      await page.evaluate(() =>
        wp.data.dispatch('core/edit-post').toggleFeature('welcomeGuide')
      );
    }

    await expect(page).toMatch('Publishing Flow Test');

    expect(await getEditedPostContent()).toMatchSnapshot();

    const werePrePublishChecksEnabled = await arePrePublishChecksEnabled();

    //eslint-disable-next-line jest/no-if
    if (werePrePublishChecksEnabled) {
      await disablePrePublishChecks();
    }

    await publishPostWithPrePublishChecksDisabled();
    await enablePrePublishChecks();

    const postPermalink = await page.evaluate(() =>
      wp.data.select('core/editor').getPermalink()
    );

    await Promise.all([page.goto(postPermalink), page.waitForNavigation()]);

    await expect(page).toMatch('Publishing Flow Test');
    await expect(page).toMatchElement('amp-story-player');
  });
});
