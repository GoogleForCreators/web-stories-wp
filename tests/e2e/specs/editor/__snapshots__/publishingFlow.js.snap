// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Publishing Flow should guide me towards creating a new post to embed my story 1`] = `
"<!-- wp:web-stories/embed {\\"url\\":\\"http://localhost:8899/stories/publishing-flow-test\\",\\"title\\":\\"Publishing Flow Test\\",\\"poster\\":\\"\\"} -->
<div class=\\"wp-block-web-stories-embed alignnone\\"><amp-story-player style=\\"width:360px;height:600px\\" data-testid=\\"amp-story-player\\"><a href=\\"http://localhost:8899/stories/publishing-flow-test\\">Publishing Flow Test</a></amp-story-player></div>
<!-- /wp:web-stories/embed -->"
`;
